core/src/muxing.rs

```diff
@@ -231,511 +114,92 @@ pub enum StreamMuxerEvent<T> {
-            + Send
-            + Sync,
-    >,
-}
-
-impl StreamMuxerBox {
-    /// Turns a stream muxer into a `StreamMuxerBox`.
-    pub fn new<T>(muxer: T) -> StreamMuxerBox
-    where
-        T: StreamMuxer + Send + Sync + 'static,
-        T::OutboundSubstream: Send,
-        T::Substream: Send,
-    {
-        let wrap = Wrap {
-            inner: muxer,
-            substreams: Mutex::new(Default::default()),
-            next_substream: AtomicUsize::new(0),
-            outbound: Mutex::new(Default::default()),
-            next_outbound: AtomicUsize::new(0),
-        };
-
-        StreamMuxerBox {
-            inner: Box::new(wrap),
+            StreamMuxerEvent::OutboundSubstream(stream, id) => {
+                StreamMuxerEvent::OutboundSubstream(map(stream), id)
+            }
+            StreamMuxerEvent::AddressChange(addr) => StreamMuxerEvent::AddressChange(addr),
         }
     }
 }
 
-impl StreamMuxer for StreamMuxerBox {
-    type Substream = usize; // TODO: use a newtype
-    type OutboundSubstream = usize; // TODO: use a newtype
-    type Error = io::Error;
-
-    #[inline]
-    fn poll_event(
-        &self,
-        cx: &mut Context<'_>,
-    ) -> Poll<Result<StreamMuxerEvent<Self::Substream>, Self::Error>> {
-        self.inner.poll_event(cx)
-    }
-
-    #[inline]
-    fn open_outbound(&self) -> Self::OutboundSubstream {
-        self.inner.open_outbound()
-    }
-
-    #[inline]
-    fn poll_outbound(
-        &self,
-        cx: &mut Context<'_>,
-        s: &mut Self::OutboundSubstream,
-    ) -> Poll<Result<Self::Substream, Self::Error>> {
-        self.inner.poll_outbound(cx, s)
-    }
-
-    #[inline]
-    fn destroy_outbound(&self, substream: Self::OutboundSubstream) {
-        self.inner.destroy_outbound(substream)
-    }
-
-    #[inline]
-    fn read_substream(
-        &self,
-        cx: &mut Context<'_>,
-        s: &mut Self::Substream,
-        buf: &mut [u8],
-    ) -> Poll<Result<usize, Self::Error>> {
-        self.inner.read_substream(cx, s, buf)
-    }
-
-    #[inline]
-    fn write_substream(
-        &self,
-        cx: &mut Context<'_>,
-        s: &mut Self::Substream,
-        buf: &[u8],
-    ) -> Poll<Result<usize, Self::Error>> {
-        self.inner.write_substream(cx, s, buf)
-    }
+#[derive(Default, Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub struct OutboundSubstreamId(u64);
 
-    #[inline]
-    fn flush_substream(
-        &self,
-        cx: &mut Context<'_>,
-        s: &mut Self::Substream,
-    ) -> Poll<Result<(), Self::Error>> {
-        self.inner.flush_substream(cx, s)
-    }
+impl OutboundSubstreamId {
+    pub(crate) const ZERO: OutboundSubstreamId = OutboundSubstreamId(0);
 
-    #[inline]
-    fn shutdown_substream(
-        &self,
-        cx: &mut Context<'_>,
-        s: &mut Self::Substream,
-    ) -> Poll<Result<(), Self::Error>> {
-        self.inner.shutdown_substream(cx, s)
-    }
+    fn fetch_and_increment(&mut self) -> Self {
```

Comment:
```suggestion
    fn get_then_increment(&mut self) -> Self {
```

Would be more intuitive to me. Not sure whether the current version is more idiomatic?

Reply:
Naming was inspired by `Atomic::fetch_add`: https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU8.html#method.fetch_add

Could probably just call it `fetch_add` directly?

Reply:
Sounds good to me.